.ci/magician/cmd/create_test_failure_ticket.go

```diff
@@ -0,0 +1,480 @@
+			gaRate, gaRateLabel := testFailureRate(gaTestFailuresMap[tName])
+			tFailure.GaFailureRate = gaRate
+			tFailure.GaFailureRateLabel = gaRateLabel
+		}
+		if _, ok := betaTestFailuresMap[tName]; ok {
+			betaRate, betaRateLabel := testFailureRate(betaTestFailuresMap[tName])
+			tFailure.BetaFailureRate = betaRate
+			tFailure.BetaFailureRateLabel = betaRateLabel
+		}
+	}
+
+	// Get existing GitHub test failure issues
+	existTestNames, err := failingTestNamesFromActiveIssues(ctx, gh)
+	if err != nil {
+		return fmt.Errorf("error getting active test failure issues: %w", err)
+	}
+
+	// Get Github test failue issues closed (fixed) today
+	closedTestNames, err := failingTestNamesFromClosedIssuesToday(ctx, gh, now)
+	if err != nil {
+		return fmt.Errorf("error getting today's closed test failure issues: %w", err)
+	}
+
+	// Create tickets
+	for _, testFailure := range testFailuresToday {
+		if shouldCreateTicket(*testFailure, existTestNames, closedTestNames) {
+			err := createTicket(ctx, gh, testFailure)
+			if err != nil {
+				return fmt.Errorf("error creating test failure ticket: %w", err)
+			}
+			fmt.Println("Yes, should create ticket")
+		} else {
+			fmt.Println("No, should not create ticket")
+		}
+	}
+	return nil
+}
+
+func lastNDaysTestFailureMap(pVersion provider.Version, n int, now time.Time, gcs CloudstorageClient, testFailuresMap map[string][]bool, testFailuresToday map[string]*testFailure) error {
+	for i := 0; i < n; i++ {
+		date := now.AddDate(0, 0, -i)
+		testInfoList, err := getTestInfoList(pVersion, date, gcs)
+		if err != nil {
+			return fmt.Errorf("error getting test info list: %w", err)
+		}
+		for _, testInfo := range testInfoList {
+			testName := testInfo.Name
+			if _, ok := testFailuresMap[testName]; !ok {
+				testFailuresMap[testName] = make([]bool, TotalDays)
+			}
+			testFailuresMap[testName][i] = testInfo.Status == "FAILURE"
+
+			if i == 0 && testInfo.Status == "FAILURE" {
+				if _, ok := testFailuresToday[testName]; !ok {
+					testFailuresToday[testName] = &testFailure{
+						TestName:         testName,
+						AffectedResource: convertTestNameToResource(testName),
+						ErrorMessage:     testInfo.ErrorMessage,
+						DebugLogLink:     testInfo.LogLink,
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func testFailureRate(testFailures []bool) (string, testFailureRateLabel) {
+	n := len(testFailures)
+	failCount := 0
+	last3DaysFailed := true
+
+	for i, fail := range testFailures {
+		if fail {
+			failCount++
+		} else if i < 3 {
+			last3DaysFailed = false
+		}
+	}
+
+	// test passed consistently for last n days
+	if failCount == 0 {
+		return "0%", testFailureNone
+	}
+
+	// test failed consistently for last 3 days
+	if last3DaysFailed {
+		return "100%", testFailure100
+	}
+
+	failRate := float64(failCount) / float64(n)
+	failRateStr := fmt.Sprintf("%.0f%%", failRate*100)
+
+	switch {
+	case failRate >= 0.5:
+		return failRateStr, testFailure50
+	case failRate >= 0.1:
```

Comment:
I'm a little concerned that this might have false positives - if it's been passing and only the previous night failed, it'll have a 7-day failure rate of 14% and will trigger this case (getting marked as 10% even if it ends up being 100%. But I may be missing something. Maybe we could just skip labeling the 10% for now?

Reply:
We're not currently creating tickets for failures with a lower than 50% rate, so we won't trigger the false positive test case at this time. If we decide to include lower failure rate cases under the current system, I'm thinking using a longer period, like30 days, for < 50% and implementing logic to manage ticket statuses to failure rate labels accordingly.

Reply:
great, thanks for confirming!